### PR TITLE
fix custom buttons

### PR DESCRIPTION
**Pull request by Skalii**

### CODE REVIEW

**Review by tigrim — APPROVED** [review]

File: Sources/Helpers/OASvgHelper.mm [hunk 1]
```diff
@@ -15,6 +15,11 @@
 
 @implementation OASvgHelper
 
++ (BOOL)hasMapImageNamed:(NSString *)name
+{
+    return name && ([self mapImagePathFromSvgResource:[name hasPrefix:@"mx_"] ? name : [@"mx_" stringByAppendingString:name]]) != nil;
```
Comment (alex-osm) [inline Sources/Helpers/OASvgHelper.mm:20]: Map image cannot have mx_ images. This request should be explained or completely reworked.

Comment (alex-osm) [inline Sources/Helpers/OASvgHelper.mm:20]: We have also "c_mx_" images. Such common method should take it into consideration.

File: Sources/POI/OAPOIUIFilter.mm [hunk 2]
```diff
@@ -1071,12 +1072,20 @@ - (void) setTypeToAccept:(OAPOICategory *)poiCategory b:(BOOL)b
...
+    NSString *iconName;
     if ([filterId hasPrefix:STD_PREFIX])
-        return standardIconId;
+        iconName = standardIconId;
     else if ([filterId hasPrefix:USER_PREFIX])
-        return [[filterId substringFromIndex:USER_PREFIX.length] lowerCase];
-
-    return filterId;
+        iconName = [[filterId substringFromIndex:USER_PREFIX.length] lowerCase];
+    if ([OASvgHelper hasMapImageNamed:iconName])
```
Comment (alex-osm) [inline Sources/POI/OAPOIUIFilter.mm:1080]: Not good idea to check if file exists at this method. Whole fix should be discussed.

File: Sources/Helpers/OASvgHelper.mm [hunk 3]
```diff
@@ -10,11 +10,21 @@
 #import "OANativeUtilities.h"
 
 #include <OsmAndCore/SkiaUtilities.h>
+#include <QReadWriteLock>
 
 const static float kDefaultIconSize = 24.0f;
 
+static NSMutableDictionary<NSString *, NSString *> *resourcesExistCache = [NSMutableDictionary dictionary];
```
Comment (alex-osm) [inline Sources/Helpers/OASvgHelper.mm:17]: It is better to have single cache `resourcesPaths` with empty string if resource does not exist.

File: Sources/Helpers/OASvgHelper.mm [hunk 4]
```diff
@@ -10,11 +10,21 @@
...
+static NSMutableArray<NSString *> *resourcesNotExistCache = [NSMutableArray array];
+static QReadWriteLock iconsCacheLock;
```
Comment (alex-osm) [inline Sources/Helpers/OASvgHelper.mm:19]: `resourcesPathsLock`

File: Sources/Helpers/OASvgHelper.mm [hunk 5]
```diff
@@ -29,37 +39,72 @@ + (nullable UIImage *) mapImageNamed:(NSString *)name scale:(float)scale
...
+            }
+            resourcesExistCache[resourceName] = resourcePath;
+        }
+    }
+    float scaledSize = kDefaultIconSize * UIScreen.mainScreen.scale;
     return [OANativeUtilities skImageToUIImage:[OANativeUtilities skImageFromSvgResourcePath:resourcePath width:scaledSize height:scaledSize]];
 }
 
-+ (UIImage *) mapImageFromSvgResource:(NSString *)resourceName width:(float)width height:(float)height
++ (nullable NSString *)mapImagePathFromSvgResource:(NSString *)resourceName
```
Comment (alex-osm) [inline Sources/Helpers/OASvgHelper.mm:71]: Create single method for operating with cache like `(nullable NSString *)mapImageResourcePathNamed:(NSString *)resourceName resourceDir:(NSString *)resourceDir`

**Pull request merged**